tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc

```diff
@@ -366,40 +520,65 @@ NVPTXCompiler::CompileTargetBinary(const HloModuleConfig& module_config,
   std::pair<int, int> compute_capability =
       absl::get<std::pair<int, int>>(gpu_version);
 
-  std::string libdevice_dir;
-  {
-    tensorflow::mutex_lock lock(mutex_);
+  bool use_cache = persistent_compilation_cache_.InUse();
+  bool have_ptx = false;
+  int64 key;
+  string ptx;
+  if (use_cache) {
+    key = persistent_compilation_cache_.CreateKey(
+      llvm_module, compute_capability, PtxOptsFromConfig(module_config));
+    have_ptx = persistent_compilation_cache_.LookupCache(key, ptx);
+  }
 
-    // Find the directory containing libdevice.  To avoid searching for it every
-    // time, we have a one-element cache, keyed on the module's config's
-    // cuda_data_dir.
-    if (cached_libdevice_dir_.empty()) {
-      cached_libdevice_dir_ = GetLibdeviceDir(module_config);
+  if (!have_ptx) {
+    std::string libdevice_dir;
+    {
+      tensorflow::mutex_lock lock(mutex_);
+
+      // Find the directory containing libdevice. To avoid searching for it
+      // every time, we have a one-element cache, keyed on the module's
+      // config's cuda_data_dir.
+      if (cached_libdevice_dir_.empty()) {
+	cached_libdevice_dir_ = GetLibdeviceDir(module_config);
+      }
+      libdevice_dir = cached_libdevice_dir_;
+    }
+    VLOG(2) << "Libdevice dir = " << libdevice_dir << "\n";
+    std::unique_ptr<llvm::Module> loaded_module =
+        MaybeLoadLLVMFromFile(debug_module, llvm_module);
+    llvm::Module *selected_module = nullptr;
+    if (loaded_module) {
+      selected_module = loaded_module.get();
+    } else {
+      selected_module = llvm_module;
     }
-    libdevice_dir = cached_libdevice_dir_;
-  }
-  VLOG(2) << "Libdevice dir = " << libdevice_dir << "\n";
-  std::unique_ptr<llvm::Module> loaded_module =
-      MaybeLoadLLVMFromFile(debug_module, llvm_module);
-  llvm::Module* selected_module = nullptr;
-  if (loaded_module) {
-    selected_module = loaded_module.get();
-  } else {
-    selected_module = llvm_module;
-  }
 
-  string ptx;
-  if (!(debug_module &&
-        MaybeLoadPtxFromFile(module_config, debug_module, &ptx))) {
-    XLA_SCOPED_LOGGING_TIMER(
-        "NVPTXCompiler::CompileTargetBinary - CompileToPtx");
-    TF_ASSIGN_OR_RETURN(ptx, nvptx::CompileToPtx(selected_module, gpu_version,
-                                                 module_config, libdevice_dir));
+    if (!(debug_module &&
+          MaybeLoadPtxFromFile(module_config, debug_module, &ptx))) {
+      XLA_SCOPED_LOGGING_TIMER(
+          "NVPTXCompiler::CompileTargetBinary - CompileToPtx");
+      TF_ASSIGN_OR_RETURN(ptx,
+                          nvptx::CompileToPtx(selected_module, gpu_version,
+                                              module_config, libdevice_dir));
+      if (use_cache) {
+	persistent_compilation_cache_.AddToCache(key, ptx);
+      }
+    }
```

Comment:
The `!have_ptx` bit could be extracted into a separate function.

Reply:
I changed the code a little, but did not make it a function.'The function would be very specific to a single call site.